### PR TITLE
exponentially back off while retrying on getting pod annotations

### PR DESCRIPTION
**Pull request by rajatchopra**
Signed-off-by: Rajat Chopra <rchopra@redhat.com>

Testing on a large cluster conveys that we should not pound the api server so heavily (or at least so frequently). This PR implements the backoff strategy.

### CODE REVIEW

**Comment by rajatchopra** [comment]
@shettyg PTAL

File: go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go [hunk 1]
```diff
@@ -152,26 +153,27 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+	// Exponential back off ~32 seconds + 7* t(api call)
+	var annotationBackoff = wait.Backoff{Duration: 1 * time.Second, Steps: 7, Factor: 1.5, Jitter: 0.1}
 	var annotation map[string]string
-	for cnt := 0; cnt < 30; cnt++ {
+	if err := wait.ExponentialBackoff(annotationBackoff, func() (bool, error) {
 		annotation, err = kubecli.GetAnnotationsOnPod(namespace, podName)
 		if err != nil {
-			time.Sleep(1 * time.Second)
-			continue
+			// maybe check if err is non recoverable
```
Comment (shettyg) [inline go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go]: The comment confused me a bit. Instead of "maybe", can you write XXX?

**Comment by shettyg** [comment]
One comment. If tested on couple of nodes, LGTM. 

We use similar logic in pods.go. Not sure whether we want to do something similar there.

**Comment by rajatchopra** [comment]
@shettyg Fixed the comment.

We do not need to fix the loop in pods.go because that happens only at the master in a sequential fashion. The one in the cni (running on the node) requires a fix - as I found by creating 6000 pods in one shot. All of them across all nodes starting to trouble the api server and brought it down. True DDOS attack :).

Nevertheless I have changed it so that it uses the same exponential backoff style.

File: go-controller/pkg/ovn/pods.go [hunk 2]
```diff
@@ -140,23 +141,22 @@ func (oc *Controller) deleteLogicalPort(pod *kapi.Pod) {
...
+	if err := wait.ExponentialBackoff(getPodBackoff, func() (bool, error) {
 		p, err := oc.Kube.GetPod(pod.Namespace, pod.Name)
 		if err != nil {
-			logrus.Errorf("Could not get pod %s/%s for obtaining the logical switch it belongs to", pod.Namespace, pod.Name)
-			continue
+			logrus.Warningf("Could not get pod %s/%s for obtaining the logical switch it belongs to", pod.Namespace, pod.Name)
+			return false, nil
 		}
 		logicalSwitch = p.Spec.NodeName
+		return true, nil
```
Comment (shettyg) [inline go-controller/pkg/ovn/pods.go]: I think the logic has changed now from before. If logicalSwitch is empty, we will return as "true" and the retry will fail.

Comment (shettyg) [reply go-controller/pkg/ovn/pods.go]: It may make sense to revert the change from pods.go, since it does not contribute with DDOS.

**Comment by rajatchopra** [comment]
@shettyg I understood what you are saying about the pods.go change. It is perhaps more suitable to keep that incremental loop on there, we don't need exponential backoff there yet.

Removed that commit.

**Comment by shettyg** [comment]
LGTM (if tested)

**Comment by rajatchopra** [comment]
Test ok.

**Pull request merged**